src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SkeletonReferenceCache.cs

```diff
@@ -0,0 +1,354 @@
+        /// yield if something else is computing and will then use the single instance computed once one thread succeeds.
+        /// </summary>
+        private readonly SemaphoreSlim _emitGate = new(initialCount: 1);
+
+        /// <summary>
+        /// Lock around <see cref="_version"/> and <see cref="_skeletonReferenceSet"/> to ensure they are updated/read 
+        /// in an atomic fashion.
+        /// </summary>
+        private readonly object _stateGate = new();
+
+        /// <summary>
+        /// The <see cref="Project.GetDependentSemanticVersionAsync"/> version of the project that the
+        /// <see cref="_skeletonReferenceSet"/> corresponds to.
+        /// </summary>
+        private VersionStamp? _version;
+
+        /// <summary>
+        /// Mapping from metadata-reference-properties to the actual metadata reference for them.
+        /// </summary>
+        private SkeletonReferenceSet? _skeletonReferenceSet;
+
+        public SkeletonReferenceCache()
+            : this(version: null, skeletonReferenceSet: null)
+        {
+        }
+
+        private SkeletonReferenceCache(
+            VersionStamp? version,
+            SkeletonReferenceSet? skeletonReferenceSet)
+        {
+            _version = version;
+            _skeletonReferenceSet = skeletonReferenceSet;
+        }
+
+        /// <summary>
+        /// Produces a copy of the <see cref="SkeletonReferenceCache"/>, allowing forks of <see cref="ProjectState"/> to
+        /// reuse <see cref="MetadataReference"/>s when their dependent semantic version matches ours.  In the case where
+        /// the version is different, then the clone will attempt to make a new skeleton reference for that version.  If it
+        /// succeeds, it will use that.  If it fails however, it can still use our skeletons.
+        /// </summary>
+        public SkeletonReferenceCache Clone()
+        {
+            lock (_stateGate)
+            {
+                // pass along the best version/reference-set we computed for ourselves.  That way future ProjectStates
+                // can use this data if either the version changed, or they weren't able to build a skeleton for themselves.
+                // By passing along a copy we ensure that if they have a different version, they'll end up producing a new
+                // SkeletonReferenceSet where they'll store their own data in which will not affect prior ProjectStates.
+                return new SkeletonReferenceCache(_version, _skeletonReferenceSet);
+            }
+        }
+
+        public async Task<MetadataReference?> GetOrBuildReferenceAsync(
+            ICompilationTracker compilationTracker,
+            SolutionState solution,
+            MetadataReferenceProperties properties,
+            CancellationToken cancellationToken)
+        {
+            // First, just see if we have cached a reference set that is complimentary with the version of the project
```

Comment:
this comment should probably be moved to where `TryReadSkeletonReferenceSetAtThisVersion` is called

Reply:
done.